### PR TITLE
GitHub Actions workflow to publish specification to GitHub Pages

### LINKED ISSUE #145: Automated publishing workflow
Automate publishing the specification to GitHub pages via Github Actions.

My current proposal is:
1. a CI workflow will build each PR and publish to an appropriately named directory in the `gh-pages` branch, i.e. this PR might be published to `wip/pulls/143` (and then would be browsable at theupdateframework.github.io/specifcation/wip/pulls/143). The workflow would also update the PR with a link to the rendered document.
2. a cleanup workflow would run periodically and compare the children of `wip/pulls` to open PRs, removing any children of `wip/pulls` for which there isn't a corresponding open PR
3. a release workflow will be triggered on changes to `master`, and:
    1. make a release on GitHub with a corresponding versioned tag
    2. build the spec for that release and publish it to the `latest` directory of the `gh-pages` branch
    3. also publish the built spec to a versioned subdirectory of the `gh-pages` branch
    4. update a listing page (index.html?) to point to the versioned subdirectory (the listing will enable browsers to find: the latest spec, specific versioned spec releases, and the most recent draft)
    5. __NOTE__: once this is done we need to update theupdateframework.io to point to the generated latest specification (and index of published versions)
4. a draft workflow will be triggered on changes to `draft`, and:
    1. build the spec in the `draft` branch
    2. publish the spec to a `draft` subdirectory of the `gh-pages` branch
5. a nag workflow will be triggered on changes to `master` and file an issue requesting `draft` is updated

Questions:
* Should we be signing spec releases (and therefore not doing automated releases on push to master)?
* Is the cleanup workflow necessary or, for the sake of a ~200K per PR, should we retain history for the PRs?

_Originally posted by @joshuagl in https://github.com/theupdateframework/specification/issues/143#issuecomment-762334990_

- [ ] PRs trigger a build and publish to a wip/pulls subdirectory of the `gh-pages` branch
- [ ] Updated master branch triggers a release (tag in repo and published release on GitHub)
- [ ] Release triggers a build and publishes to _both_ latest and versioned subdirectories of `gh-pages` branch (and updates index)
- [ ] theupdateframework.io updated to link to published specification on theupdateframework.github.io/specification
- [ ] Updated master branch files an issue to rebase draft branch
- [ ] Updated draft branch builds the spec and publishes it to the draft subdirectory of the `gh-pages` branch

### PR DESCRIPTION
**Pull request by joshuagl**
Here's a simple collection of scripts and GitHub Actions workflows to build and publish the bikeshed specification when it is updated.

On pushes to master the workflow will:
* create a release
* build that release
* publish it to both a versioned subdirectory of the gh-pages branch _and_ the latest subdirectory of the gh-pages branch
* build a simple HTML index to point to all published versions of the bikshed specification (and the draft version)

On pushes to draft the workflow will:
* build the draft
* publish it to a draft subdirectory of the gh-pages branch

You can see an example of the automation here in action at https://joshuagl.github.io/specification/

**Note**: we need GitHub Actions and GitHub Pages enabling for the repository for this to work.

Fixes #145

### CODE REVIEW

**Comment by joshuagl** [comment]
You can see an example of this GHA workflow in action on my personal fork here https://github.com/joshuagl/specification/pull/5/checks?check_run_id=1870772911

**Comment by joshuagl** [comment]
Also an example of a failed run, because date was not changed, here: https://github.com/joshuagl/specification/pull/6/checks?check_run_id=1903153676

**Comment by joshuagl** [comment]
I've updated this PR to not only convert to GitHub Actions, but also to include workflows to build, release (in the case of master) & publish the specification on pushes to the master and draft branches.

Please take a look @lukpueh, @trishankatdatadog and @mnm678.

File: .github/workflows/pr.yml
Comment (lukpueh) [inline .github/workflows/pr.yml:7]: We only run the release checker here, right? Is this name accurate?

File: .github/workflows/pr.yml
Comment (lukpueh) [inline .github/workflows/pr.yml:3]: f9fd90af5ad8233ccab54baf8a61a35e9134d311 says "we will configure CI to only run check_release on PRs to master". Do we have to configure this here?

**Review by lukpueh — COMMENTED** [review]
Very impressive work, @joshuagl, and as usual, very nicely laid out (re commit separation and messages)!! 🎉 

I'd like to take another look at the release and draft workflow files with a fresh set of eyes on Monday, but on a quick glance they seem to do what they should.

File: build_index.py
Comment (joshuagl) [reply build_index.py:180]: Good question! I think it does not make sense to enforce the sanity check. Our GHA workflow ensures we are on the correct branch, and we should let folks using the script offline decide where to run it. Indeed, I had to comment this out for testing and completely forgot I had even written it...

File: .github/workflows/pr.yml
Comment (joshuagl) [reply .github/workflows/pr.yml:7]: Nice catch, I will update the name.

File: .github/workflows/pr.yml
Comment (joshuagl) [reply .github/workflows/pr.yml:3]: We do! I will fix that, thank you.

File: get_version.py
Comment (joshuagl) [reply get_version.py:19]: That's right, I wanted to ensure we could replicate the GHA workflow outside of on an Actions runner with minimal effort.

**Comment by joshuagl** [comment]
Thank you for the detailed review @lukpueh. I've addressed your comments by removing the sanity check code, renaming the PR workflow name and adding a branch specifier to the PR workflow.

File: check_release.py
Comment (lukpueh) [inline check_release.py:145]: > don't forget to tag the release ..

Not that this ever really reminded someone, but it can definitely go now. :)

**Review by lukpueh — APPROVED** [review]
LGTM modulo some minor nits that you may address or not. Many thanks for the excellent PR, @joshuagl!

File: .github/workflows/draft.yml
Comment (joshuagl) [reply .github/workflows/draft.yml:47]: I like the idea of using the mailing list!

File: .github/workflows/draft.yml
Comment (joshuagl) [reply .github/workflows/draft.yml:5]: That's right, a merge pushes changes to the branch and results in a 'push' event.

File: .github/workflows/release.yml
Comment (joshuagl) [reply .github/workflows/release.yml]: Yes, that would be cool. Not much work at all, I've added it in 948022f

File: check_release.py
Comment (joshuagl) [reply check_release.py:144]: Removed in 47869c0

**Comment by joshuagl** [comment]
> LGTM modulo some minor nits that you may address or not. Many thanks for the excellent PR, @joshuagl!

Some good nits, I've addressed each in some follow-on commits. Please review the extra 3 patches and re-approve the PR at your convenience.

**Review by lukpueh — APPROVED** [review]
💯

**Comment by joshuagl** [comment]
If anyone has the time, I would appreciate a second review on this so we can get it merged and start publishing the newly formatted specification.

**Review by mnm678 — APPROVED** [review]
Looks great, thanks for all the work on this!

**Pull request merged**
